src/shelly_hap_temperature_sensor.cpp

```diff
@@ -0,0 +1,152 @@
+    : Component(id),
+      Service(SHELLY_HAP_IID_BASE_TEMPERATURE_SENSOR +
+                  SHELLY_HAP_IID_STEP_SENSOR * (id - 1),
+              &kHAPServiceType_TemperatureSensor,
+              kHAPServiceDebugDescription_TemperatureSensor),
+      temp_sensor_(std::move(sensor)),
+      cfg_(cfg) {
+}
+
+TemperatureSensor::~TemperatureSensor() {
+}
+
+Component::Type TemperatureSensor::type() const {
+  return Type::kSensor;
+}
+
+Component::SensorType TemperatureSensor::sensor_type() const {
+  return SensorType::kTemperatureSensor;
+}
+
+std::string TemperatureSensor::name() const {
+  return cfg_->name;
+}
+
+Status TemperatureSensor::SetConfig(const std::string &config_json,
+                                    bool *restart_required) {
+  struct mgos_config_se cfg = *cfg_;
+  cfg.name = nullptr;
+  json_scanf(config_json.c_str(), config_json.size(), "{name: %Q, unit: %d",
+             &cfg.name, &cfg.unit);
+
+  mgos::ScopedCPtr name_owner((void *) cfg.name);
+  // Validation.
+  if (cfg.name != nullptr && strlen(cfg.name) > 64) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s",
+                        "name (too long, max 64)");
+  }
+  if (cfg.unit < 0 || cfg.unit > 1) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid unit");
+  }
+  // Now copy over.
+  if (cfg_->name != nullptr && strcmp(cfg_->name, cfg.name) != 0) {
+    mgos_conf_set_str(&cfg_->name, cfg.name);
+    *restart_required = true;
+  }
+  if (cfg_->unit != cfg.unit) {
+    cfg_->unit = cfg.unit;
+  }
+  return Status::OK();
+}
+
+Status TemperatureSensor::SetState(const std::string &state_json) {
+  (void) state_json;
+  return Status::OK();
+}
+
+void TemperatureSensor::ValueChanged() {
+  current_temperature_characteristic->RaiseEvent();
+}
+
+Status TemperatureSensor::Init() {
+  uint16_t iid = svc_.iid + 1;
+  current_temperature_characteristic = new mgos::hap::FloatCharacteristic(
+      iid++, &kHAPCharacteristicType_CurrentTemperature, 0.0, 100.0, 0.1,
+      std::bind(&TemperatureSensor::HandleCurrentTemperatureRead, this, _1, _2,
+                _3),
+      true /* supports_notification */, nullptr /* write_handler */,
+      kHAPCharacteristicDebugDescription_CurrentTemperature);
+  AddChar(current_temperature_characteristic);
+  AddChar(new mgos::hap::UInt8Characteristic(
+      iid++, &kHAPCharacteristicType_TemperatureDisplayUnits, 0, 255, 1,
+      std::bind(&TemperatureSensor::HandleTemperatureDisplayUnitsRead, this, _1,
+                _2, _3),
+      true /* supports_notification */,
+      std::bind(&TemperatureSensor::HandleTemperatureDisplayUnitsWrite, this,
+                _1, _2, _3),
+      kHAPCharacteristicDebugDescription_TemperatureDisplayUnits));
+  temp_sensor_->notifier_ = std::bind(&TemperatureSensor::ValueChanged, this);
+  LOG(LL_INFO, ("Exporting Temp"));
+
+  temp_sensor_->StartUpdating(5000);
+  return Status::OK();
+}
+
+HAPError TemperatureSensor::HandleCurrentTemperatureRead(
```

Comment:
you can remove some boilerplate by converting these to lambda functions, see here - https://github.com/mongoose-os-apps/shelly-homekit/commit/fdeac316a6baf2c2010fefff3c63c8c23d60d986